internal/impl/packages.go

```diff
@@ -0,0 +1,344 @@
+		}
+		fmt.Printf("%s\n", msg)
+
+		var cmd *exec.Cmd
+		if pkg != "" {
+			cmd = exec.Command(
+				"nix", "profile", "install",
+				"--profile", profileDir,
+				"--extra-experimental-features", "nix-command flakes",
+				nix.FlakeNixpkgs(d.cfg.Nixpkgs.Commit)+"#"+pkg,
+			)
+		} else {
+			cmd = exec.Command(
+				"nix", "flake", "prefetch",
+				"--extra-experimental-features", "nix-command flakes",
+				nix.FlakeNixpkgs(d.cfg.Nixpkgs.Commit),
+			)
+		}
+
+		cmd.Env = nix.DefaultEnv()
+		cmd.Stdout = &nixPackageInstallWriter{d.writer}
+		cmd.Stderr = cmd.Stdout
+		err = cmd.Run()
+		var exitErr *exec.ExitError
+		if errors.As(err, &exitErr) {
+			msg := fmt.Sprintf(
+				"running command %s: exit status %d with command stderr: %s",
+				cmd, exitErr.ExitCode(), string(exitErr.Stderr),
+			)
+			color.New(color.FgRed).Printf("%s: Fail\n", msg)
+			return errors.New(msg)
+		}
+		if err != nil {
+			msg := fmt.Sprintf("running command %s: %v", cmd, err)
+			color.New(color.FgRed).Printf("%s: Fail\n", msg)
+			return errors.New(msg)
+		}
+		fmt.Printf("%s: ", msg)
+		color.New(color.FgGreen).Printf("Success\n")
+	}
+
+	return nil
+}
+
+func (d *Devbox) removePackagesFromProfile(pkgs []string) error {
+	if !featureflag.Flakes.Enabled() {
+		return nil
+	}
+
+	items, err := d.nixProfileListItems()
+	if err != nil {
+		return err
+	}
+
+	nameToAttributePath := map[string]string{}
+	for _, item := range items {
+		attrPath, err := item.attributePath()
+		if err != nil {
+			return err
+		}
+		packageName, err := item.packageName()
+		if err != nil {
+			return err
+		}
+		nameToAttributePath[packageName] = attrPath
+	}
+
+	profileDir, err := d.profileDir()
+	if err != nil {
+		return err
+	}
+
+	for _, pkg := range pkgs {
+		attrPath, ok := nameToAttributePath[pkg]
+		if !ok {
+			return errors.Errorf("Did not find attributePath for package: %s", pkg)
+		}
+
+		cmd := exec.Command("nix", "profile", "remove",
+			"--profile", profileDir,
+			"--extra-experimental-features", "nix-command flakes",
+			attrPath,
+		)
+		cmd.Stdout = d.writer
+		cmd.Stderr = d.writer
+		err = cmd.Run()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *Devbox) pendingPackagesForInstallation() ([]string, error) {
+	if featureflag.Flakes.Disabled() {
+		return nil, errors.New("Not implemented for legacy non-flakes devbox")
+	}
+
+	items, err := d.nixProfileListItems()
+	if err != nil {
+		return nil, err
+	}
+
+	installed := map[string]bool{}
+	for _, item := range items {
+		packageName, err := item.packageName()
+		if err != nil {
+			return nil, err
+		}
+		installed[packageName] = true
+	}
+
+	pending := []string{}
+	for _, pkg := range d.cfg.Packages {
+		if _, ok := installed[pkg]; !ok {
+			pending = append(pending, pkg)
+		}
+	}
+	return pending, nil
+}
+
+// nixProfileListItems returns a list of the installed packages
+func (d *Devbox) nixProfileListItems() ([]*nixProfileListItem, error) {
```

Comment:
I think this and below code is better moved to `nix/profile.go` so it'll be `nix.ProfileListItems()`